### PR TITLE
2306 contract api sr originate

### LINKED ISSUE #2306: [Mumbai] Add Smart_rollup_originate support in the Contract API
Add support for the new operation `smart_rollup_originate` in these following providers:
- RpcContractProvider
- EstimateProvider
- PrepareProvider

Additional requirements:
- Update types/interfaces as necessary
- Add unit and integration tests
- Update documentation

Example of smart_rollup_originate:
```
{
  "protocol": "PtMumbaiiFFEGbew1rRjzSPyzRbA51Tm3RVZL5suHPxSZYDhCEc",
  "chain_id": "NetXQw6nWSnrJ5t",
  "hash": "onfaRBp9oLvaJ3E72LkvgzoNjsyz8bpBt9jGwVZdnDPmGejTFvY",
  "branch": "BLPA6B27rQ82TQiPoGGzjqwS8fvFfb9E1Beq3t5QdiVPtQeJV3J",
  "contents": [
    {
      "kind": "reveal",
      "source": "tz2Q3yRaczTqZVf3ZQvwiiTqKjhJFyDzeRSz",
      "fee": "359",
      "counter": "12054",
      "gas_limit": "1000",
      "storage_limit": "0",
      "public_key": "sppk7d2tnVN58p2fnxD2ru52kZs71YxdLv4CpWSp1DhkwdaE1vtUeTJ",
      "metadata": {
        "balance_updates": [
          {
            "kind": "contract",
            "contract": "tz2Q3yRaczTqZVf3ZQvwiiTqKjhJFyDzeRSz",
            "change": "-359",
            "origin": "block"
          },
          {
            "kind": "accumulator",
            "category": "block fees",
            "change": "359",
            "origin": "block"
          }
        ],
        "operation_result": {
          "status": "applied",
          "consumed_milligas": "1000000"
        }
      }
    },
    {
      "kind": "smart_rollup_originate",
      "source": "tz2Q3yRaczTqZVf3ZQvwiiTqKjhJFyDzeRSz",
      "fee": "957",
      "counter": "12055",
      "gas_limit": "2849",
      "storage_limit": "6572",
      "pvm_kind": "wasm_2_0_0",
      "kernel": "0061736d0100000001280760037f7f7f017f60027f7f017f60057f7f7f7f7f017f60017f0060017f017f60027f7f0060000002610311736d6172745f726f6c6c75705f636f72650a726561645f696e707574000011736d6172745f726f6c6c75705f636f72650c77726974655f6f7574707574000111736d6172745f726f6c6c75705f636f72650b73746f72655f77726974650002030504030405060503010001071402036d656d02000a6b65726e656c5f72756e00060aa401042a01027f41fa002f0100210120002f010021022001200247044041e4004112410041e400410010021a0b0b0800200041c4006b0b5001057f41fe002d0000210341fc002f0100210220002d0000210420002f0100210520011004210620042003460440200041016a200141016b10011a0520052002460440200041076a200610011a0b0b0b1d01017f41dc0141840241901c100021004184022000100541840210030b0b38050041e4000b122f6b65726e656c2f656e762f7265626f6f740041f8000b0200010041fa000b0200020041fc000b0200000041fe000b0101",
      "origination_proof": "0300020c4a316fa1079bfc23dac5ecc609ab10e26490e378a81e774c51176040bea180467070f4682a44b982768d522ec6380982f446488c0176ed7c13aa1d6c12a03a810764757261626c658108726561646f6e6c79d00b749948da9186d29aed2f9327b46793f18b1e6499c40f0ddbf0bf785e85e2e9",
      "parameters_ty": {
        "prim": "bytes"
      },
      "metadata": {
        "balance_updates": [
          {
            "kind": "contract",
            "contract": "tz2Q3yRaczTqZVf3ZQvwiiTqKjhJFyDzeRSz",
            "change": "-957",
            "origin": "block"
          },
          {
            "kind": "accumulator",
            "category": "block fees",
            "change": "957",
            "origin": "block"
          }
        ],
        "operation_result": {
          "status": "applied",
          "balance_updates": [
            {
              "kind": "contract",
              "contract": "tz2Q3yRaczTqZVf3ZQvwiiTqKjhJFyDzeRSz",
              "change": "-1638000",
              "origin": "block"
            },
            {
              "kind": "burned",
              "category": "storage fees",
              "change": "1638000",
              "origin": "block"
            }
          ],
          "address": "sr1HxZ24YmT3hkpeSaxb6N7vrpp6BfbPSrEh",
          "genesis_commitment_hash": "src135NQY1CGVUaYdE5Jp4PZn8CzGuznh4bz8z5cwtuCbSanLMdPER",
          "consumed_milligas": "2748269",
          "size": "6552"
        }
      }
    }
  ],
  "signature": "sigRjjHRNcKkfHxRf7kMeQ1Qsjn4SzX6e7Y4zKZqYHDiWC4C9yHDg5f4pusM6usfbitBxmAS9efcVUxBg8TpD6MhR5fipAya"
}
```

### PR DESCRIPTION
**Pull request by zainen**
add support for `smart_rollup_originate` operation 

closes #2306 
Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [x] Your code builds cleanly without any errors or warnings
- [x] You have run the linter against the changes
- [x] You have added unit tests (if relevant/appropriate)
- [x] You have added integration tests (if relevant/appropriate)
- [x] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

In this PR, please also make sure: 

- [ ] You have linked this PR to the issue by putting `closes #TICKETNUMBER` in the description box (when applicable)
- [ ] You have added a concise description on your changes
## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

File: packages/taquito/src/operations/smart-rollup-originate-operation.ts
Comment (hui-an-yang) [inline packages/taquito/src/operations/smart-rollup-originate-operation.ts:18]: small thing, update ConsensusKey to SmartRollupOriginate in the beginning and the end of the sentence

File: packages/taquito/src/operations/operation-emitter.ts
Comment (hui-an-yang) [inline packages/taquito/src/operations/operation-emitter.ts:186]: Thought getFee function is taking care of storage_limit too

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (hui-an-yang) [inline packages/taquito/src/estimate/rpc-estimate-provider.ts:472]: 🙌good catch!

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (hui-an-yang) [inline packages/taquito/src/estimate/rpc-estimate-provider.ts:178]: Can you elaborate a bit what does line 178-182 do?

File: packages/taquito-rpc/src/types.ts
Comment (hui-an-yang) [inline packages/taquito-rpc/src/types.ts:483]: Is `pvm_kind` that we want to create enum like the api doc? To prevent user from sending any string and got rejected by chain?

File: packages/taquito/src/prepare/prepare-provider.ts
Comment (hui-an-yang) [inline packages/taquito/src/prepare/prepare-provider.ts:229]: same here, thought this.getFee took care of storage_limit too

File: packages/taquito/test/contract/rpc-contract-provider.spec.ts
Comment (hui-an-yang) [inline packages/taquito/test/contract/rpc-contract-provider.spec.ts:145]: Will update with a smartRollupOriginate test suite when available, right?

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (zainen) [reply packages/taquito/src/estimate/rpc-estimate-provider.ts:178]: removing casting
ts was complaining but i think i needed to refresh my ide since its not complaining anymore

File: packages/taquito/src/operations/operation-emitter.ts
Comment (zainen) [reply packages/taquito/src/operations/operation-emitter.ts:186]:
```
const getFee = (op: RPCOpWithFee) => {
      const opCounter = ++counters[publicKeyHash];
      return {
        counter: `${opCounter}`,
        fee: typeof op.fee === 'undefined' ? '0' : `${op.fee}`,
        gas_limit: typeof op.gas_limit === 'undefined' ? '0' : `${op.gas_limit}`,
        storage_limit: typeof op.storage_limit === 'undefined' ? '0' : `${op.storage_limit}`,
      };
    };
```
have to dig deeper but this made the difference because for som reason its still being seen as undefined

File: packages/taquito/test/contract/rpc-contract-provider.spec.ts
Comment (zainen) [reply packages/taquito/test/contract/rpc-contract-provider.spec.ts:145]: thanks for the catch thought i added a test already

File: integration-tests/contract-smart-rollup-originate.spec.ts
Comment (hui-an-yang) [inline integration-tests/contract-smart-rollup-originate.spec.ts:3]: small clean of ", "

File: packages/taquito-rpc/src/rpc-client-interface.ts
Comment (hui-an-yang) [inline packages/taquito-rpc/src/rpc-client-interface.ts:131]: With this rpc endpoint added, might want to add a test case in `integration-tests/rpc-nodes.spec.ts`

Comment (zainen) [reply packages/taquito-rpc/src/rpc-client-interface.ts:131]: will do thanks :D

File: packages/taquito/src/operations/smart-rollup-originate-operation.ts
Comment (hui-an-yang) [inline packages/taquito/src/operations/smart-rollup-originate-operation.ts:18]: last word to be SmartRollupOriginate

File: integration-tests/contract-smart-rollup-originate.spec.ts
Comment (hui-an-yang) [inline integration-tests/contract-smart-rollup-originate.spec.ts:16]: mondayAndAlpha here cuz they didn't support originationProof endpoint yet?

File: packages/taquito/src/operations/smart-rollup-originate-operation.ts
Comment (hui-an-yang) [inline packages/taquito/src/operations/smart-rollup-originate-operation.ts:18]: might want to update last word to be SmartRollupOriginate instead of UpdateConsensusKey

File: packages/taquito/src/operations/smart-rollup-originate-operation.ts
Comment (hui-an-yang) [reply packages/taquito/src/operations/smart-rollup-originate-operation.ts]: sorry not being clear enough, I was trying to say update last word to be SmartRollupOriginate instead of UpdateConsensusKey

File: packages/taquito-rpc/src/taquito-rpc.ts
Comment (hui-an-yang) [inline packages/taquito-rpc/src/taquito-rpc.ts:1179]: might want to be consistent using "params" defined in function instead of "reqs" contains ...?

File: integration-tests/contract-smart-rollup-originate.spec.ts
Comment (dsawali) [inline integration-tests/contract-smart-rollup-originate.spec.ts:4]: missing semicolons in this file

File: integration-tests/contract-smart-rollup-originate.spec.ts
Comment (dsawali) [inline integration-tests/contract-smart-rollup-originate.spec.ts:26]: I don't think this comment is really necessary, it should already be implied

File: packages/taquito-contracts-library/src/rpc-wrapper.ts
Comment (dsawali) [inline packages/taquito-contracts-library/src/rpc-wrapper.ts]: could we rename this? reqs doesn't seem like it has anything to do with what the value actually is

File: packages/taquito/src/contract/rpc-contract-provider.ts
Comment (dsawali) [inline packages/taquito/src/contract/rpc-contract-provider.ts:741]: extraneous line break/white space

File: packages/taquito/src/contract/rpc-contract-provider.ts
Comment (dsawali) [inline packages/taquito/src/contract/rpc-contract-provider.ts:774]: extraneous line break, please right click 'format document' on file to fix these

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (dsawali) [inline packages/taquito/src/estimate/rpc-estimate-provider.ts:173]: kind of a nitpick, but since we're in the scope of forEach for `result`, we should leverage the values inside that iterated value

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (dsawali) [inline packages/taquito/src/estimate/rpc-estimate-provider.ts:738]: typo

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (dsawali) [inline packages/taquito/src/estimate/rpc-estimate-provider.ts:753]: I think we can just pass the object, no need to make a new variable

File: packages/taquito/test/contract/rpc-contract-provider.spec.ts
Comment (dsawali) [inline packages/taquito/test/contract/rpc-contract-provider.spec.ts:2332]: please follow the existing test patterns of checking the `opOb` being outputted, and keep in mind the 3 cases:
- op with reveal
- op without reveal
- op with fees (or other estimates) overridden

File: packages/taquito/test/prepare/prepare-provider.spec.ts
Comment (dsawali) [inline packages/taquito/test/prepare/prepare-provider.spec.ts:1078]: please add 1 more case with the reveal

**Review by dsawali — CHANGES_REQUESTED** [review]

**Comment by dsawali** [comment]
@zainen missing batch support and tests for batching

**Comment by zainen** [comment]
> @zainen missing batch support and tests for batching

ahh ok right I'll add that

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (zainen) [reply packages/taquito/src/estimate/rpc-estimate-provider.ts:173]: oh thanks for hte comment i started working on this before which should take in a constant but was put on hold until  the other pr was merged

Comment (zainen) [reply packages/taquito/src/estimate/rpc-estimate-provider.ts]: this has changed but comment still applies 
the check is for content which operationResults does not contain and unsure if there are every conditions where tehre are multiple values inside of the operation restults that could cause for the storage limit to be added multiple times

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (dsawali) [inline packages/taquito/src/estimate/rpc-estimate-provider.ts:171]: we should move in this condition inside the loop as discussed

File: packages/taquito/src/operations/operation-errors.ts
Comment (dsawali) [inline packages/taquito/src/operations/operation-errors.ts:78]: why partial?

**Review by dsawali — CHANGES_REQUESTED** [review]

File: packages/taquito/src/operations/operation-errors.ts
Comment (zainen) [reply packages/taquito/src/operations/operation-errors.ts:78]: because otherwise it is expected that other operations have items that are specific to smartRollupOriginate

File: packages/taquito/test/contract/rpc-contract-provider.spec.ts
Comment (zainen) [reply packages/taquito/test/contract/rpc-contract-provider.spec.ts:2332]: hmmm i definitely added this but dont see it 👀 i'll add it again

File: packages/taquito/test/prepare/prepare-provider.spec.ts
Comment (zainen) [reply packages/taquito/test/prepare/prepare-provider.spec.ts:1078]: same here

File: packages/taquito/src/operations/operation-errors.ts
Comment (zainen) [reply packages/taquito/src/operations/operation-errors.ts:78]: otherwise the other types are expected to have the same expected key item pairs

File: packages/taquito/test/contract/rpc-contract-provider.spec.ts
Comment (zainen) [reply packages/taquito/test/contract/rpc-contract-provider.spec.ts:2332]: not sure what happened to my commit that fixed this 😦

File: docs/rpc_package.md
Comment (dsawali) [inline docs/rpc_package.md:118]: nitpick, but comment is a bit redundant

File: integration-tests/contract-smart-rollup-originate.spec.ts
Comment (dsawali) [inline integration-tests/contract-smart-rollup-originate.spec.ts:9]: this should be above the describe block

File: integration-tests/contract-smart-rollup-originate.spec.ts
Comment (dsawali) [inline integration-tests/contract-smart-rollup-originate.spec.ts:21]: extraneous new line

File: integration-tests/injecting-multiple-manager-operations-in-a-block.spec.ts
Comment (dsawali) [inline integration-tests/injecting-multiple-manager-operations-in-a-block.spec.ts:4]: unnecessary trailing comma

File: packages/taquito-contracts-library/src/read-provider-wrapper.ts
Comment (dsawali) [inline packages/taquito-contracts-library/src/read-provider-wrapper.ts:49]: is this now a proto constant?

File: packages/taquito-rpc/src/taquito-rpc.ts
Comment (dsawali) [inline packages/taquito-rpc/src/taquito-rpc.ts:1179]: I think it's a bit more clear to say 'PVM kind' rather than 'kind (PVM)', we could also ditch the 'from' at the end

File: packages/taquito-rpc/src/taquito-rpc.ts
Comment (dsawali) [inline packages/taquito-rpc/src/taquito-rpc.ts:1180]: nitpick,

for a* smart rollup originate op

I think the verbage 'the' can be misconstrued by users

File: packages/taquito/src/contract/rpc-contract-provider.ts
Comment (dsawali) [inline packages/taquito/src/contract/rpc-contract-provider.ts:739]: typo

**Review by dsawali — CHANGES_REQUESTED** [review]

File: packages/taquito/src/estimate/rpc-estimate-provider.ts
Comment (zainen) [reply packages/taquito/src/estimate/rpc-estimate-provider.ts:170]: ahh ok i didnt realize that the `genesis_commitment_hash` was sr specific

File: packages/taquito-contracts-library/src/read-provider-wrapper.ts
Comment (zainen) [reply packages/taquito-contracts-library/src/read-provider-wrapper.ts:49]: yes but based on the estimate its a bit more complicated to try to estimate it ourself and its held in the estimate return within `result.size`

File: packages/taquito/src/batch/rpc-batch-provider.ts
Comment (zainen) [reply packages/taquito/src/batch/rpc-batch-provider.ts:202]: this follows the convention of the other batch operations minus the missed 'the'

`Add an operation to submit a tx rollup batch to the batch`

Comment (zainen) [reply packages/taquito/src/batch/rpc-batch-provider.ts:202]: should this be updated for all batch operations in favour the new wording? or an issue created to do it?

Comment (dsawali) [reply packages/taquito/src/batch/rpc-batch-provider.ts:202]: the tx rollup batch doesn't make much sense, if you don't mind updating that in this PR that would be great

File: packages/taquito-contracts-library/src/read-provider-wrapper.ts
Comment (zainen) [reply packages/taquito-contracts-library/src/read-provider-wrapper.ts:49]: do you want this constant removed if it currently isnt in use?

**Review by dsawali — APPROVED** [review]

**Pull request merged**
